source/ed/osm2ed.cpp

```diff
@@ -204,44 +213,103 @@ void ReadNodesVisitor::node_callback(uint64_t osm_id, double lon, double lat, co
+            Admins* context = reinterpret_cast<Admins*>(c);
             context->push_back(rel);
         }
         return true;
     };
     admin_tree.Search(search_rect.min, search_rect.max, callback, &result);
-    for (auto rel : result) {
-        if (boost::geometry::within(p, rel->polygon)) {
+    for (const auto* rel : result) {
+        if (boost::geometry::covered_by(p, rel->polygon)) {
             return rel;
         }
     }
+    if (this->cities_db) {
+        return find_admin_in_cities(lon, lat);
+    }
     return nullptr;
 }
 
+template <typename T>
+T read_wkt(const std::string& s) {
+    T g;
+    boost::geometry::read_wkt(s, g);
+    return g;
+}
+
+// This method searches for admin around the coord in the 'cities' database
+// It does a bulk search for all the admin in a bounding box around the coord
+// to reduce the number of calls to the datase
+const Admin* OSMCache::find_admin_in_cities(const double lon, const double lat) {
+    const auto p = point(lon, lat);
+    std::stringstream request;
+    request << "SELECT id, uri, name, coalesce(insee, '') as insee, level, coalesce(post_code, '') as post_code, "
+            << "ST_ASTEXT(coord) as center, "
+            << "ST_ASTEXT(boundary) as boundary "
+            << "FROM administrative_regions "
+            << "WHERE ST_Intersects(ST_MakeEnvelope(" << std::setprecision(16) << lon - cities_bbox_to_fetch << ", "
+            << lat - cities_bbox_to_fetch << ", " << lon + cities_bbox_to_fetch << ", " << lat + cities_bbox_to_fetch
+            << ", " << 4326 << "), boundary)";
+
+    this->cities_db_calls++;
+
+    pqxx::work work(*cities_db);
+    pqxx::result result = work.exec(request);
+
+    const Admin* containing_admin = nullptr;
+    for (const auto& it : result) {
+        auto polygon = read_wkt<mpolygon_type>(it["boundary"].as<std::string>());
+        auto center = read_wkt<point>(it["center"].as<std::string>());
+        const auto id = it["id"].as<u_int64_t>();
+
+        if (this->admins.find(id) != this->admins.end()) {
+            continue;
+        }
+        this->admin_from_cities++;
+        auto admin = std::make_unique<Admin>(id, it["uri"].as<std::string>(), it["insee"].as<std::string>(),
+                                             it["post_code"].as<std::string>(), it["name"].as<std::string>(),
+                                             it["level"].as<uint32_t>(), std::move(polygon), std::move(center));
+
+        if (boost::geometry::covered_by(p, admin->polygon)) {
+            containing_admin = admin.get();
+        }
+
+        // we also store the admin for future use
+        boost::geometry::model::box<point> box;
+        boost::geometry::envelope(admin->polygon, box);
+        Rect r(box.min_corner().get<0>(), box.min_corner().get<1>(), box.max_corner().get<0>(),
+               box.max_corner().get<1>());
+        this->admin_tree.Insert(r.min, r.max, admin.get());
```

Comment:
It feels weird to create a `std::unique_ptr` but still copying the pointer twice with `this->admin_tree.Insert(r.min, r.max, admin.get())` and `containing_admin = admin.get()`.

Having a `unique_ptr` that shares its pointer sounds like a `std::shared_ptr` to me !

Reply:
hum I don't think so, `shared_ptr` is for shared ownership, but here the ownership is only for the main container